### ISSUE TITLE
Devise a better system for "re-releasing"

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#1864]
Right now, release builds are triggered by creating a release on GitHub. If the build fails, the only way to "re-run" is to create a new release. So we have a few versions that have just never been released. [v0.0.221](https://github.com/charliermarsh/ruff/actions/runs/3916906363/jobs/6696361593) is one example, where I had to push a fix, then re-release as v0.0.222.

**Comment by charliermarsh** [comment]
Actually, for that build, I'm going to try deleting the release + tag, and then re-creating it with the updated `main`.

**Comment by not-my-profile** [comment]
(Unrelated sidenote ... I don't think we need to bump the version for `ruff_dev` at all since it's just internal tooling.)

**Comment by bilderbuchi** [comment]
Depending on your desired workflow, you could insert an additional `release_prep` between `release` and the others it `uses` (moving those to `release_prep`).
Then, you could fire the `release_prep` either on PRs that have a `release` (or other appropriate) label, or on pushes to a `release` branch, or any other condition that you like.
When that check is successful, you can push a release commit with git tag & Github release, and the full machinery should run as it does now.

**Comment by stinodego** [comment]
Creating a release also creates a tag. You can have your workflow trigger on a new tag. If anything goes wrong, you only need to update the tag, which will re-trigger the workflow. I think that solves your original problem.

**Comment by bilderbuchi** [comment]
Hm, I always create and push the git tag first (manually), and then the Github release.
Also, updating the tag would mean rewriting public history, no? I guess it depends on the project if that is OK.

**Comment by stinodego** [comment]
I figure updating a tag is fine as long as there is no associated release on PyPI yet.

**Comment by charliermarsh** [comment]
Ah yeah, interesting. So the release would create a tag, which _could_ fail; then I could fix whatever bug caused it to fail; then I could push the tag again to the newer SHA, and the job would run again, and succeed. Is that right?

**Comment by stinodego** [comment]
Exactly.

So you would change the release workflow trigger to:

```yaml
on:
  push:
    tags:
      - '**'
```

This will be triggered when you do a manual GitHub release, as it creates a new tag.

Then, if the workflow fails, you add a new commit with a fix, and you update the tag like so:

```
git tag -f <tag-name> <commit-hash> && git push -f origin <tag-name>
```

This will re-trigger the release workflow and now (hopefully) successfully release to PyPI.

### Limitations

Note that the new commit, and any commits you merge before that fix, will not automatically be added to the release changelog of the existing release. But they will be part of the release.

After updating the tag, you'd have to manually edit the existing release to re-generate the changelog. The auto-generate should pick up the new tags just fine and generate the correct changelog, so this is no issue at all.

**Comment by charliermarsh** [comment]
I used the above strategy once, it's a bit manual but it works great if needed. I've also started to run a release build prior to creating the release which catches non-transient issues.
